### PR TITLE
Be more focused in the build

### PR DESCRIPTION
**Pull request by tjmw**
## What does this change?

Before:

* All tests etc for all workspaces run multiple times - once per subproject in the matrix
* When the tests failed for a particular workspace, it wasn't the package step that failed but the "Upload to Riff-Raff" step, which feels a bit accidental (only failed because there's no zip file). See [this run](https://github.com/guardian/support-service-lambdas/actions/runs/9582507082/job/26421631106) as an example.

After:

* Only run the tests etc for the current subproject in the matrix (and the cdk project, which we always need)
* When the tests fail, the `package` step fails, so we fail faster and it's clearer where the issue is

## Accessibility

-   [ ] [Tested with screen reader](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#screen-reader)
-   [ ] [Navigable with keyboard](https://github.com/guardian/accessibility/blob/main/people-and-technology/02-physical.md#keyboard)
-   [ ] [Colour contrast passed](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#contrast)
-   [ ] [The change doesn't use only colour to convey meaning](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#use-of-colour)

### CODE REVIEW

File: _templates/new-lambda/api-gateway/package.ejs.t [hunk 1]
```diff
@@ -12,7 +12,7 @@ sh: git add handlers/<%=lambdaName%>/package.json
     "type-check": "tsc --noEmit",
     "build": "esbuild --bundle --platform=node --target=node18 --outfile=target/index.js src/index.ts",
     "lint": "eslint src/**/*.ts",
-    "package": "pnpm type-check && pnpm lint && pnpm test && pnpm build && cd target; zip -qr <%=lambdaName%>.zip ./*.js",
+    "package": "pnpm type-check && pnpm lint && pnpm test && pnpm build && cd target && zip -qr <%=lambdaName%>.zip ./*.js",
```
Comment (tjmw) [inline _templates/new-lambda/api-gateway/package.ejs.t:15]: Previously if the `pnpm test` command failed, the overall exit code for `package` was 0. I guess this must be becuase the `zip` command was somehow succeeding. This resulted in the `package` step in the GH action succeeding which is a bit misleading. Then we'd fail later on in the upload to Riff Raff step.

File: .github/workflows/ci-typescript.yml [hunk 2]
```diff
@@ -40,7 +40,8 @@ jobs:
           cache: 'pnpm'
       - run: pnpm install
       - run: pushd modules && pnpm test; popd
```
Comment (johnduffell) [inline .github/workflows/ci-typescript.yml:42]: > Only run the tests etc for the current subproject in the matrix (and the cdk project, which we always need)

we will also run all the tests for all the modules

**Review by johnduffell — COMMENTED** [review]
can you check that the type-check/lint will check the modules still, as before it was relying ont he top level `package` command in the package.json to search out all the ts files bc eslint or tsc.
We might have to add an extra step to do this similar to the way we run the tests.

File: modules/product-catalog/src/api.ts [hunk 3]
```diff
@@ -14,6 +14,7 @@ export const getProductCatalogFromApi = async (
 	});
 
 	const json = await response.json();
+	return 1;
```
Comment (rupertbates) [inline modules/product-catalog/src/api.ts:17]: Was this an accidental commit?

Comment (tjmw) [reply modules/product-catalog/src/api.ts:17]: Ha yes, I wanted to understand how TS errors in `modules` were being surfaced (see John's comment above). I'll revert it now...

File: .github/workflows/ci-typescript.yml (hunk 2 above)
Comment (tjmw) [reply .github/workflows/ci-typescript.yml:42]: 👍  Yeah - I think that's okay. Maybe as a future alternative we could run the module tests in a separate step and fail early if they're broken.

**Review by johnduffell — APPROVED** [review]
turns out linting is (already) not running on the modules, so we're not losing anything, thanks for making it better Tom 👍

**Comment by tjmw** [comment]
> turns out linting is (already) not running on the modules, so we're not losing anything, thanks for making it better Tom 👍

Thanks John. Yep linting of modules doesn't currently happen in CI (either in main or this branch) - one for a future PR I think. Type checking of modules does happen, via the consuming handler. This means that if a module is used by two handlers, both will fail the TS check - I think that's okay. It also means if a module is defined but not used by anything it won't be type checked in CI.

**Review by rupertbates — APPROVED** [review]
This is great work, thanks for picking it up 👏👏👏

File: .github/workflows/ci-typescript.yml (hunk 2 above)
Comment (tjmw) [reply .github/workflows/ci-typescript.yml:43]: @johnduffell I've changed this in #2359 where I've introduced a separate job for the common stuff (module tests, module formatting)

**Pull request merged**